File: app/src/main/java/com/example/muzik/response_model/Album.java
package com.example.muzik.response_model;

import android.net.Uri;

import java.util.List;

public class Album implements ResponseModel {
    private long albumID;
    private String name;
    private Uri albumArtURI;
    private String imageURL;
    private long artistID;
    private String albumArtist;

    private List<Song> songs;

    public Album(boolean isNewSample) {
       if(isNewSample) albumID = -1;
    }

    public Album(long albumID, String name, String albumArtist, Uri albumArtURI, List<Song> songs) {
        this.albumID = albumID;
        this.name = name;
        this.albumArtURI = albumArtURI;
        this.songs = songs;
    }

    public Album(long albumID, String name, String imageURL, long artistID, List<Song> songs) {
        this.albumID = albumID;
        this.name = name;
        this.imageURL = imageURL;
        this.artistID = artistID;
        this.songs = songs;
    }

    public long getAlbumID() {
        return albumID;
    }

    public String getAlbumArtist() {
        return albumArtist;
    }

    public void setAlbumID(long albumID) {
        this.albumID = albumID;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getImageURL() {
        return imageURL;
    }

    public void setImageURL(String imageURL) {
        this.imageURL = imageURL;
    }

    public long getArtistID() {
        return artistID;
    }

    public void setArtistID(long artistID) {
        this.artistID = artistID;
    }

    public List<Song> getSongs() {
        return songs;
    }

    public void setSongs(List<Song> songs) {
        this.songs = songs;
    }

    public Uri getAlbumArtURI() {
        return albumArtURI;
    }

    public void setAlbumArtURI(Uri albumArtURI) {
        this.albumArtURI = albumArtURI;
    }
}

File: app/src/main/java/com/example/muzik/data_model/standard_model/Album.kt
package com.example.muzik.data_model.standard_model

import android.net.Uri
import com.example.muzik.data_model.api_model.response.AlbumResponse
import com.example.muzik.music_service.LocalMusicRepository

class Album (
    val name: String? = null,
    val imageURI: Uri? = null,
    val albumID: Long? = -1,
    val artistID: Long? = -1,
    val artistName: String? = ""
) : Model {
    val listSongId: MutableList<Long> = ArrayList()

    companion object {
        fun buildLocal(albumID: Long, album: String, albumArtUri: Uri, artistID: Long): Album {
            return Album(
                albumID = albumID,
                name = album,
                imageURI = albumArtUri,
                artistID = artistID
            )
        }

        fun buildOnline(albumResponse: AlbumResponse): Album {
            return Album(
                albumID = albumResponse.albumID,
                name = albumResponse.name,
                imageURI = Uri.parse(albumResponse.imageURL),
                artistID = albumResponse.artistID,
                artistName = albumResponse.artistName
            )
        }
    }

    fun getArtist(): Artist {
        return artistID?.let { LocalMusicRepository.getArtist(it) } as Artist
    }

    fun getSongs(): List<Song> {
        val listSong: MutableList<Song> = ArrayList()
        for (songId: Long in listSongId) {
            LocalMusicRepository.getSong(songId)?.let { listSong.add(it) }
        }
        return listSong
    }

    fun requireAlbumID(): Long {
        checkNotNull(albumID) {
            ("Album $this must have non-null album ID")
        }
        return albumID
    }
}